### PR TITLE
Rewrote this with ES2015 stuff

### PR DESCRIPTION
**Pull request by cag**
This is kinda an extensive changeset so of course please review this.

### CODE REVIEW

File: content/en/chapter12-writing-a-plugin.md [hunk 1]
```diff
@@ -49,30 +49,32 @@ So here’s our skeleton `index.js` file:
...
+    // 2. Precompile the marker regexp from config
+  }
 
-// Helper: escapes any regex-special character
-function escapeRegex(str) {
-  return String(str).replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
+  // On-the-fly compilation callback (file by file); assumes Brunch already
+  // accepted that file for our plugin by checking `type`, `extension` and
+  // `pattern`.
+  compile(params, callback) {
```
Comment (shvaikalesh) [inline content/en/chapter12-writing-a-plugin.md]: `Promise`

Comment (cag) [reply content/en/chapter12-writing-a-plugin.md]: Brunch has a way of using promises now?

Comment (shvaikalesh) [reply content/en/chapter12-writing-a-plugin.md]: Yep, and it will be cool if this guide use the new API.

**Comment by paulmillr** [comment]
Thanks!

**Pull request merged**